sw/device/lib/dif/dif_flash_ctrl.h

```diff
@@ -0,0 +1,1019 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 1.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_DIF_DIF_FLASH_CTRL_H_
+#define OPENTITAN_SW_DEVICE_LIB_DIF_DIF_FLASH_CTRL_H_
+
+/**
+ * @file
+ * @brief <a href="/hw/ip/flash_ctrl/doc/">flash controller</a> Device Interface
+ * Functions
+ */
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_warn_unused_result.h"
+
+// Header Extern Gaurd (so header can be used from C and C++)
+#ifdef __cplusplus
+extern "C" {
+#endif  // _cplusplus
+
+/**
+ * Configuration for initializing the flash controller device.
+ */
+typedef struct dif_flash_ctrl_config {
+  /**
+   * The base address for registers in the flash controller device.
+   */
+  mmio_region_t base_addr;
+} dif_flash_ctrl_config_t;
+
+/**
+ * Carries state for a flash controller device.
+ *
+ * All members should be considered private and should not be accessed directly
+ * outside of the DIF implementation. This struct is only public for allocation
+ * purposes.
+ */
+typedef struct dif_flash_ctrl {
+  /**
+   * The base address for registers in the flash controller device.
+   */
+  mmio_region_t base_addr;
+
+  /**
+   * The number of words transfered from the start of the latest program/read
+   * transaciton.
+   */
+  uint32_t words_transfered;
```

Comment:
In DV, I've found it useful to wrap all flash operation specific items in a single struct. 
https://github.com/lowRISC/opentitan/blob/91e1b769eba7a8c4faae396aacc8eb20eab77a85/hw/ip/flash_ctrl/dv/env/flash_ctrl_env_pkg.sv#L80